Add QC check for multiple gene associations

twhetzel opened the pull request:
closes #8316 

This PR adds a qc check and creates an error if any Mondo class has more than one gene association added either as an equivalentTo restriction or a subClassOf axiom. It is ok if the class has one of each, as long as it's the same gene (I can add that qc check).

This PR is related to[ point 7](https://github.com/monarch-initiative/mondo/pull/8108#issuecomment-2322123779) in the omim pipeline PR.

NOTE: Based on https://github.com/monarch-initiative/mondo/issues/8276 the classes MONDO:0011974 'retinitis pigmentosa 7 and MONDO:0100050 'Usher syndrome, type 1D/F' should be excluded from this QC check.
Update:  MONDO:0100188 'combined ApoA-I and ApoC-III deficiency' should also be excluded based on https://github.com/monarch-initiative/mondo/issues/8329#issuecomment-2472001053

The qc failures in this PR are handled in https://github.com/monarch-initiative/mondo/pull/8330 (edits on a newer version of mondo-edit.obo).

matentzn commented on src/sparql/qc/mondo/qc-gene-identifier-mismatch.sparql:
Very nice test. I like it a lot, hopefully it is not too slow, but looks ok, given that it failed after the usual 10 minutes.

The only minor feedback that I have is that to make things easier for the future, I tend to now only `SELECT DISTINCT ?entity ?property ?value` - this way, we can easily integrate the test into a dashboard and ROBOT report if we so wisheth. However, I don't know if that will ever happen, and granted this way of listing all properties like you have it is more readable. So, I leave this choice in your hands.

twhetzel replied:
Thanks for explaining why `SELECT DISTINCT ?entity ?property ?value` is usually the way the qc checks are written. I was aiming for something more readable here so let's leave this one as-is and if needed can trim items out of the SELECT later.

matentzn commented on src/sparql/qc/mondo/qc-gene-identifier-mismatch.sparql:
```diff
@@ -0,0 +1,35 @@
+
+  # Filter for gene identifiers with HGNC or NCBIGene prefixes
+  FILTER(STRSTARTS(STR(?equivGeneIdentifier), "http://identifiers.org/hgnc/") || 
+         STRSTARTS(STR(?equivGeneIdentifier), "http://identifiers.org/ncbigene/"))
```
I assume you checked that they are `http` vs `https`.

twhetzel replied:
Thanks for asking. Yes, I checked it is `http`. The terms to import now come in through this [gSheet](https://docs.google.com/spreadsheets/d/1UME3pTeR42hwNt1I6RPl0nvJKVfTIeJqhx80laAlD50/edit?gid=0#gid=0) that we pull from to add into `manual_seed.txt` when refreshing imports and there are also already qc failures of this check that we are resolving in another [ticket](https://github.com/monarch-initiative/mondo/issues/8329).

matentzn commented on src/sparql/qc/mondo/qc-multiple-gene-associations.sparql:
```diff
@@ -0,0 +1,34 @@
+               owl:onProperty obo:RO_0004003 ;
+               owl:someValuesFrom ?geneIdentifier .
+  }
+}
```
My feeling is that in a UNION query, the bindings are renewed, so ?entity in the first union does not have to be bound to the same ?entity in the second union query. To mitigate this, you can move the ?label outside the UNION, like so:

```
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX obo: <http://purl.obolibrary.org/obo/>
PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>

SELECT DISTINCT ?entity ?label (GROUP_CONCAT(DISTINCT ?geneIdentifier; separator=", ") AS ?geneIdentifiers)
WHERE {
  {
    # subClassOf association
    ?entity rdfs:subClassOf ?restriction .
    ?restriction rdf:type owl:Restriction ;
                 owl:onProperty obo:RO_0004003 ;
                 owl:someValuesFrom ?geneIdentifier .
  }
  UNION
  {
    # equivalentClass association
    ?entity owl:equivalentClass ?equivClass .
    ?equivClass (owl:intersectionOf/rdf:rest*/rdf:first) ?component .
    ?component rdf:type owl:Restriction ;
               owl:onProperty obo:RO_0004003 ;
               owl:someValuesFrom ?geneIdentifier .
  }

  # Ensure consistent labeling, if available
  ?entity rdfs:label ?label .
}
GROUP BY ?entity ?label
HAVING (COUNT(DISTINCT ?geneIdentifier) > 1)
```

Now the entity is bound outside the union, and inside they will bind to the same.

Could be wrong though, just tickling the back of my head?

twhetzel replied:
I had these as separate query files before, one qc query to check if there is more than one gene association stated using the equivalentTo restriction and one qc query to check if there is more than one gene association stated as a subClassOf relationship, so I'm not trying to check if the same entity is used in both. After having each query work separately, I though you might ask me to combine them so I did... but they can also easily be separate qc queries.

twhetzel replied:
I'm not sure if we are both trying to get the query to check the same thing because I don't see why the ?entity needs to be the same for both the equivalentTo restriction and subClassOf. I want the qc check to fail if either of these ways of representing a gene association in Mondo have >1 gene association. So fail if >1 gene association represented as a subClassOf or fail if >1 gene association represented in the logical definition.

Given that, do you think the query still needs to be changed?

matentzn replied:
I think you might be right.. I retract my comment, lgtm!

twhetzel replied:
well someone needs to officially approve the PR :)

joeflack4 commented on src/sparql/qc/mondo/qc-gene-identifier-mismatch.sparql:
Just wanna confirm my understanding. I'm missing some background knowledge.

Since there is a filter for hgnc/ncbigene, I'm just wondering how the OMIM disease-gene associations make their way into Mondo. Because the `omim.owl` creates them as MIM-MIM associations. I guess then somewhere the gene MIM is mapped to hgnc or ncbigene.

matentzn replied:
You yourself have implemented this mapping: https://github.com/monarch-initiative/omim/releases/download/2024-11-10/mondo-omim-genes.robot.tsv!

joeflack4 replied:
Ah OK, I remember this. In my false memory these were SKOS mappings, not _has material basis in germline mutation in_.

twhetzel replied:
Gene associations are added into Mondo from both the OMIM gene pipeline (Nico referred to) and from curators directly. When gene associations are added by curators, they can be added for both human diseases and non-human diseases in Mondo. The non-human diseases will have identifiers from NCBI Gene. Any of these curator added gene associations that are not found in OMIM, but are requested by other collaborators will have a different `source` for the gene association so the OMIM gene pipeline will not change these when it's run.

twhetzel commented:
The current QC failures were fixed in Mondo in https://github.com/monarch-initiative/mondo/pull/8330

matentzn reviewed (approved):
LGTM!

twhetzel commented:
There are still issues with the QC check. It is currently failing after the roundtrip process and re-running the QC checks on `mondo-base.obo` as seen [here](https://github.com/monarch-initiative/mondo/actions/runs/11985167435/job/33416771382?pr=8328#step:4:114).

** FAIL Rule ../sparql/qc/mondo/qc-gene-identifier-mismatch.sparql: 2 violation(s) **
```
entity,label,equivGeneIdentifier,subClassGeneIdentifier
http://purl.obolibrary.org/obo/MONDO_0100050,"Usher syndrome, type 1D/F",http://identifiers.org/hgnc/13733,http://identifiers.org/hgnc/14674
http://purl.obolibrary.org/obo/MONDO_0100050,"Usher syndrome, type 1D/F",http://identifiers.org/hgnc/14674,http://identifiers.org/hgnc/13733
```
Looking at `mondo-base.obo` in Protege, MONDO:0100050 'Usher syndrome, type 1D/F' now has the _addition_ of gene associations as SubClassOf (previously was only equivalentTo).
![mondo-base - Usher syndrome, type 1D:F](https://github.com/user-attachments/assets/44ed3efd-7eb1-456d-a7ee-d8164f29a373)

Also, after the roundtrip, in mondo-base.obo this QC check is also failing:
** FAIL Rule ../sparql/qc/mondo/qc-multiple-gene-associations.sparql: 2 violation(s) **
```
http://purl.obolibrary.org/obo/MONDO_0100188,combined ApoA-I and ApoC-III deficiency,"http://identifiers.org/hgnc/610, http://identifiers.org/hgnc/600"
http://purl.obolibrary.org/obo/MONDO_0100050,"Usher syndrome, type 1D/F","http://identifiers.org/hgnc/14674, http://identifiers.org/hgnc/13733"
```

I'm trying to sort out these issues. Any pointers @matentzn would be appreciated 😄

matentzn commented:
Due to the `relax` command which is run on all release files in all ontologies in our parts of OBO, equivalent class axioms are broken up into subclass axioms at release time, in order to create more "graphy" ontology edges.

So you will have to either remove or modify the equivalent class statement, _or_ you can encode an exception for Usher syndrome so the query is not run?

twhetzel commented:
For MONDO:0100050 'Usher syndrome, type 1D/F' based on the [OMIM information](https://omim.org/entry/601067) and past discussions I believe it's representation of the gene association as a logical definition is correct because it is a 'digenic' disease where both mutations in both genes are required to cause the condition. I can add an exclusion for MONDO:0100050 'Usher syndrome, type 1D/F' (your second suggestion) and see if that works....

twhetzel commented:
I added the exclusion to the MONDO:0100050 'Usher syndrome, type 1D/F' and added the FILTER to the query, but in the roundtrip checks these same qc checks are still failing:
- FAIL Rule ../sparql/qc/mondo/qc-gene-identifier-mismatch.sparql: 2 violation(s)
- FAIL Rule ../sparql/qc/mondo/qc-multiple-gene-associations.sparql: 2 violation(s)

matentzn commented:
Do they both have the FILTER clause for exceptions?

twhetzel commented:
Here are the queries:
# Find classes with mismatched gene identifiers added as equivalentTo and subClassOf

SELECT DISTINCT ?entity ?label ?equivGeneIdentifier ?subClassGeneIdentifier
WHERE {
  ?entity rdf:type owl:Class ;
          rdfs:label ?label .
  
  # Equivalent class restriction
  ?entity owl:equivalentClass ?equivClass .
  ?equivClass owl:intersectionOf/rdf:rest*/rdf:first ?equivComponent .
  
  ?equivComponent rdf:type owl:Restriction ;
                  owl:onProperty obo:RO_0004003 ;
                  owl:someValuesFrom ?equivGeneIdentifier .
  
  # subClassOf restriction
  ?entity rdfs:subClassOf ?subClassRestriction .
  ?subClassRestriction rdf:type owl:Restriction ;
                       owl:onProperty obo:RO_0004003 ;
                       owl:someValuesFrom ?subClassGeneIdentifier .

  FILTER NOT EXISTS {
      ?entity mondo:excluded_from_qc_check mondoSparqlQcMondo:qc-gene-identifier-mismatch.sparql .
    }
  
  # Filter for gene identifiers with HGNC or NCBIGene prefixes
  FILTER(STRSTARTS(STR(?equivGeneIdentifier), "http://identifiers.org/hgnc/") || 
         STRSTARTS(STR(?equivGeneIdentifier), "http://identifiers.org/ncbigene/"))
  FILTER(STRSTARTS(STR(?subClassGeneIdentifier), "http://identifiers.org/hgnc/") || 
         STRSTARTS(STR(?subClassGeneIdentifier), "http://identifiers.org/ncbigene/"))

  # Filter for cases where the gene identifiers do not match
  FILTER(?equivGeneIdentifier != ?subClassGeneIdentifier)
}


# Get classes that have more than 1 gene association (either subClassOf or equivalentClass) with RO:0004003 property

SELECT DISTINCT ?entity ?label (GROUP_CONCAT(DISTINCT ?geneIdentifier; separator=", ") AS ?geneIdentifiers)
WHERE {
  {
    # subClassOf association
    ?entity rdfs:subClassOf ?restriction ;
            rdfs:label ?label .
    
    ?restriction rdf:type owl:Restriction ;
                 owl:onProperty obo:RO_0004003 ;
                 owl:someValuesFrom ?geneIdentifier .
    FILTER NOT EXISTS {
      ?entity mondo:excluded_from_qc_check mondoSparqlQcMondo:qc-multiple-gene-associations.sparql .
    }
  }
  UNION
  {
    # equivalentClass association
    ?entity owl:equivalentClass ?equivClass ;
            rdfs:label ?label .
    
    ?equivClass owl:intersectionOf/rdf:rest*/rdf:first ?component .
    
    ?component rdf:type owl:Restriction ;
               owl:onProperty obo:RO_0004003 ;
               owl:someValuesFrom ?geneIdentifier .
    
    FILTER NOT EXISTS {
      ?entity mondo:excluded_from_qc_check mondoSparqlQcMondo:qc-multiple-gene-associations.sparql .
    }
  }
}
GROUP BY ?entity ?label
HAVING (COUNT(DISTINCT ?geneIdentifier) > 1)

And each Mondo class has the exclude from qc check for these queries that they are failing.

![mondo-edit - Usher syndrome, type 1D:F - excluded from checks](https://github.com/user-attachments/assets/8f38bda9-ba43-478f-9be8-9fc6573d852e)

twhetzel commented:
The exclusions are used with `mondo-base.owl` (which is _not_ the mondo-base format used in the roundtrip):
`robot query -i mondo-base.owl -q ../sparql/qc/mondo/qc-gene-identifier-mismatch.sparql _results-mondo-base-OWL_qc-gene-identifier-mismatch-ERROR.csv`
The file `_results-mondo-base-OWL_qc-gene-identifier-mismatch-ERROR.csv` is empty as expected. 

BUT are not used or some other reason is not working when run against `mondo-base.obo` (which is the mondo-base format used in the roundtrip qc checks):
`robot query -i mondo-base.obo -q ../sparql/qc/mondo/qc-gene-identifier-mismatch.sparql _results-mondo-base_qc-gene-identifier-mismatch-ERROR.csv`
The file has the errors:
```
entity,label,equivGeneIdentifier,subClassGeneIdentifier
http://purl.obolibrary.org/obo/MONDO_0100050,"Usher syndrome, type 1D/F",http://identifiers.org/hgnc/13733,http://identifiers.org/hgnc/14674
http://purl.obolibrary.org/obo/MONDO_0100050,"Usher syndrome, type 1D/F",http://identifiers.org/hgnc/14674,http://identifiers.org/hgnc/13733
```

twhetzel commented:
@matentzn I removed the qc check from being used in the roundtrip checks on the obo format file due to the mondoBase prefix issue [here](https://github.com/monarch-initiative/mondo/pull/8328/commits/c9427861fb04f1c253379495d933c9ec1b98d370#diff-1376bbad1ef43bed6c9aed84b7d52471e63c87277b2b3f72ed0a5bdef481b638R806), but this is now failing for the same qc check on the first set of checks that use mondo-tags-reasoned.owl and none of the Mondo classes listed as errors have >1 gene association. Any ideas why this is happening?

twhetzel commented:
Thanks @matentzn.  I thought these qc checks were run against the version of mondo-edit.obo in the branch, but that clears up everything! The tech checks I've added are all working as expected now and these remaining qc errors were items to fix on the curation side.

twhetzel commented:
@sabrinatoro - the tech changes in this PR have already been reviewed and approved. There are a few edits in `mondo-edit.obo` if you would like to review and approve these.

sabrinatoro reviewed (approved):
looks good to me. 
Approving and merging.
